UDN L3 support for services

ricky-rav opened the pull request:
This PR adds UDN support for cluster IP services on L3. OVN LBs are added to the UDN-specific logical switch and gateway router.  Nodeport services are only supported if the client runs on the destination node. (For full support, ovnkube-node changes are necessary and that'll be for a separate PR)

This PR spins a service controller per network: one for the default network, as today, and one for each UDN. 

OVN External IDs have been extended for logical routers, switches and load balancers to include the network name and network role (default, primary, secondary), so that we can easily distinguish which OVN objects belong to which network.

Existing unit tests are now run also for the UDN scenario. 

E2E tests for services on UDN have been added.

kyrtapz reviewed (commented):
Thanks Riccardo!
Overall the changes look good, it seems that we followed the logic of the default svc controller which I really like :)

I will have to do another pass.

kyrtapz commented on go-controller/pkg/ovn/controller/services/services_controller.go:
```diff
@@ -637,34 +671,66 @@ func (c *Controller) onEndpointSliceDelete(obj interface{}) {
+			klog.V(5).Infof("network=%s, error=%s", c.netInfo.GetNetworkName(), err.Error())
 		} else {
-			utilruntime.HandleError(fmt.Errorf("couldn't get key for EndpointSlice %+v: %v", endpointSlice, err))
+			utilruntime.HandleError(fmt.Errorf("network=%s, couldn't get key for EndpointSlice %+v: %v", c.netInfo.GetNetworkName(), endpointSlice, err))
```
This won't happen because we are filtering out other networks from the events.

ricky-rav commented on go-controller/pkg/ovn/controller/services/services_controller.go:
```diff
@@ -650,21 +682,50 @@ func (c *Controller) queueServiceForEndpointSlice(endpointSlice *discovery.Endpo
+	if err != nil {
+		return "", err
+	}
+	return FormatKeyForService(serviceName, endpointSlice.Namespace), nil
```
YES. I was sure there was a better way of doing it, I didn't remember how exactly -.-

ricky-rav commented on go-controller/pkg/ovn/controller/services/services_controller.go:
```diff
@@ -650,21 +682,50 @@ func (c *Controller) queueServiceForEndpointSlice(endpointSlice *discovery.Endpo
 	}
 
-	c.queue.Add(key)
+	c.queue.Add(FormatKeyForService(serviceName, endpointSlice.Namespace))
```
Maybe. But if we want to use namespaced names everywhere, also skipService will take that as input and as a consequence the controller queue should also be made of namespaced names instead of strings I'm not against it, but maybe there was a reason why we're just using strings?

```
func (c *Controller) onServiceAdd(obj interface{}) {
	key, err := cache.MetaNamespaceKeyFunc(obj)
	if err != nil {
		utilruntime.HandleError(fmt.Errorf("couldn't get key for object %+v for network=%s: %v", obj, c.netInfo.GetNetworkName(), err))
		return
	}

	service := obj.(*v1.Service)
	if c.skipService(service.Name, service.Namespace) {
		return
	}
	metrics.GetConfigDurationRecorder().Start("service", service.Namespace, service.Name)
	klog.V(5).Infof("Adding service %s for network=%s", key, c.netInfo.GetNetworkName())
	c.queue.Add(key)
}

```

ricky-rav commented on go-controller/pkg/ovn/controller/services/services_controller.go:
```diff
@@ -650,21 +682,50 @@ func (c *Controller) queueServiceForEndpointSlice(endpointSlice *discovery.Endpo
+	}
+}
+
+func _getServiceNameFromEndpointSlice(endpointSlice *discovery.EndpointSlice, inDefaultNetwork bool) (string, error) {
```
Well, it's for readability that I added wrappers :)

ricky-rav commented on go-controller/pkg/ovn/base_network_controller.go:
```diff
@@ -899,3 +895,40 @@ func (bnc *BaseNetworkController) findMigratablePodIPsForSubnets(subnets []*net.
 	return ipList, nil
 }
+
+func initLoadBalancerGroups(nbClient libovsdbclient.Client, netInfo util.NetInfo) (
```
I don't know, I kept the code as it was...

ricky-rav replied:
Done.

ricky-rav commented on go-controller/pkg/ovn/secondary_layer3_network_controller.go:
```diff
@@ -296,6 +314,24 @@ func NewSecondaryLayer3NetworkController(cnci *CommonNetworkControllerInfo, netI
 	addressSetFactory := addressset.NewOvnAddressSetFactory(cnci.nbClient, ipv4Mode, ipv6Mode)
 
+	var NADLister nadlister.NetworkAttachmentDefinitionLister
+	if util.IsNetworkSegmentationSupportEnabled() {
```
Ah ops, this is wrong!!

trozet replied:
wait should this also only be started if its a primary UDN?

ricky-rav replied:
yeah, we're not supporting services for secondary networks, at least yet

tssurya replied:
@ricky-rav : go to sleep :) I know I should too... :) we'll get this in ric, don't work this late..

ricky-rav replied:
I was worried about CI, since I added several changes this evening :-D

kyrtapz replied:
> wait should this also only be started if its a primary UDN?

Yeah, we shouldn't run it for secondary network. There will be no available endpointslices for secondary services so there is no point in running the service controller.

ricky-rav commented on go-controller/pkg/ovn/controller/services/lb_config.go:
```diff
@@ -882,7 +883,9 @@ func getServicePortKey(protocol v1.Protocol, name string) string {
 // when the service needs local (per-node) endpoints, that is when ETP=local or ITP=local.
 // The node list helps to keep the resulting map small, since we're only interested in local endpoints.
-func getEndpointsForService(slices []*discovery.EndpointSlice, service *v1.Service, nodes sets.Set[string]) (map[string]lbEndpoints, map[string]map[string]lbEndpoints) {
+func getEndpointsForService(slices []*discovery.EndpointSlice, service *v1.Service, nodes sets.Set[string],
```
Yeah, it's for the log messages... maybe in the future we can have a wrapper around klog.Infof that prints the network name automatically...

ricky-rav commented on go-controller/pkg/ovn/controller/services/loadbalancer.go:
```diff
@@ -425,25 +425,49 @@ func DeleteLBs(nbClient libovsdbclient.Client, uuids []string) error {
-func getLBs(nbClient libovsdbclient.Client, allTemplates TemplateMap) ([]*LB, error) {
-	_, out, err := _getLBsCommon(nbClient, allTemplates, false)
+// getAllLBs returns a slice of load balancers found in OVN.
+func getAllLBs(nbClient libovsdbclient.Client, allTemplates TemplateMap, netInfo util.NetInfo) ([]*LB, error) {
```
To filter out what doesn't belong to the input network :)

kyrtapz replied:
doesn't getAllLBs return all load balancers?

ricky-rav replied:
true, my bad!

ricky-rav commented on go-controller/pkg/util/util.go:
```diff
@@ -527,3 +560,72 @@ func GetDefaultEndpointSlicesEventHandler(handlerFuncs cache.ResourceEventHandle
 	return eventHandler
 }
+
+func GetEndpointSlicesEventHandlerForNetwork(handlerFuncs cache.ResourceEventHandlerFuncs, netInfo NetInfo) cache.ResourceEventHandler {
```
Damn, I swear I had it in mind -.-

ricky-rav commented on go-controller/pkg/ovn/controller/services/loadbalancer.go:
```diff
@@ -425,25 +425,49 @@ func DeleteLBs(nbClient libovsdbclient.Client, uuids []string) error {
+		network, ok := externalIDs[types.NetworkExternalID]
+
+		if netInfo.IsDefault() {
+			return !ok || network == types.DefaultNetworkName
```
I wanted to be on the safe side and be backward compatible by considering no external ID as the default network

ricky-rav commented on go-controller/pkg/ovn/controller/services/loadbalancer.go:
```diff
@@ -425,25 +425,49 @@ func DeleteLBs(nbClient libovsdbclient.Client, uuids []string) error {
+func _getLBsCommon(nbClient libovsdbclient.Client, allTemplates TemplateMap, withServiceOwner bool, includeAllNetworks bool, netInfo util.NetInfo) (sets.Set[string], []*LB, error) {
+
+	// Lookup network name and network role in the OVN external IDs to check whether
+	// the OVN element with the input externalIDs belongs to this network.
```
Sure! It's to filter out anything belonging to secondary (non-primary) networks, if there are any

ricky-rav commented on go-controller/pkg/ovn/controller/services/services_controller.go:
```diff
@@ -93,6 +93,7 @@ func NewController(client clientset.Interface,
 	if err != nil {
 		return nil, err
 	}
+	klog.Infof("Done creating new service controller for network=%s", netInfo.GetNetworkName())
```
we don't need it, I must have forgotten to remove it :)

ricky-rav commented on go-controller/pkg/util/util.go:
```diff
@@ -420,6 +424,27 @@ func GetActiveNetworkForNamespace(namespace string, nadLister nadlister.NetworkA
 	return nil, &UnknownActiveNetworkError{namespace: namespace}
 }
 
+func GetNetworkRole(isPrimary bool) string {
```
Sure!

ricky-rav commented on go-controller/pkg/libovsdb/ops/loadbalancer.go:
```diff
@@ -137,3 +138,15 @@ func ListLoadBalancers(nbClient libovsdbclient.Client) ([]*nbdb.LoadBalancer, er
+// based on a given predicate
+func FindLoadBalancersWithPredicate(nbClient libovsdbclient.Client, p loadBalancerPredicate) ([]*nbdb.LoadBalancer, error) {
+	found := []*nbdb.LoadBalancer{}
+	ctx, cancel := context.WithTimeout(context.Background(), types.OVSDBTimeout)
```
yes

ricky-rav commented on go-controller/pkg/ovn/controller/services/services_controller.go:
```diff
@@ -48,16 +49,19 @@ const (
 )
 
 var NoServiceLabelError = fmt.Errorf("endpointSlice missing %s label", discovery.LabelServiceName)
+var NoUDNServiceLabelError = fmt.Errorf("endpointSlice missing %s label", types.LabelUserDefinedServiceName)
```
I should have used it in `queueServiceForEndpointSlice` :

```
func (c *Controller) queueServiceForEndpointSlice(endpointSlice *discovery.EndpointSlice) {
	serviceNamespacedName, err := c.getServiceNamespacedNameFromEndpointSlice(endpointSlice)
	if err != nil {
		// Do not log endpointsSlices missing service labels as errors.
		// Once the service label is eventually added, we will get this event
		// and re-process.
		if errors.Is(err, NoServiceLabelError) || errors.Is(err, NoUDNServiceLabelError) {
			klog.V(5).Infof("network=%s, error=%s", c.netInfo.GetNetworkName(), err.Error())
		} else {
			utilruntime.HandleError(fmt.Errorf("network=%s, couldn't get key for EndpointSlice %+v: %v", c.netInfo.GetNetworkName(), endpointSlice, err))
		}
		return
	}

```
But ok, let's remove it

ricky-rav commented on go-controller/pkg/ovn/controller/services/utils.go:
```diff
@@ -25,3 +32,28 @@ func IsHostEndpoint(endpointIP string) bool {
+func getExternalIDsForLoadBalancer(service *v1.Service, netInfo util.NetInfo) map[string]string {
+	nsn := ktypes.NamespacedName{Namespace: service.Namespace, Name: service.Name}
+
+	gk := service.GroupVersionKind().GroupKind()
```
yeah, it used an object, but it was only called on services :) I was too afraid to remove it, but in the end `gk.String()`  should be the same as `Services.Kind`. Let me try it out.

ricky-rav replied:
`Services.Kind` is always empty... not sure what it is for then

ricky-rav replied:
I'll set it to `"Service"` and that's it

ricky-rav commented on go-controller/pkg/util/util.go:
```diff
@@ -420,6 +424,27 @@ func GetActiveNetworkForNamespace(namespace string, nadLister nadlister.NetworkA
+// when on the default cluster network, for backward compatibility.
+func GenerateExternalIDsForSwitchOrRouter(netInfo NetInfo) map[string]string {
+	externalIDs := make(map[string]string)
+	if netInfo.IsSecondary() {
```
ah yes, it's time to discuss this. I could safely add external IDs to LBs that are also on the default network because secondary networks don't use services. For routers and switches I preferred not to add external IDs when on the default network, at least at first. I didn't wanna break the existing code, but let me check now what secondary networks do with them...

ricky-rav replied:
In the end we decided not to alter the LBs in the default network and only add network-related external IDs to LBs in UDNs.

trozet commented on go-controller/pkg/ovn/controller/services/services_controller.go:
```diff
@@ -544,16 +544,38 @@ func (c *Controller) RequestFullSync(nodeInfos []nodeInfo) {
-	klog.V(5).Infof("Adding service %s", key)
+
 	service := obj.(*v1.Service)
+	if c.skipService(service.Name, service.Namespace) {
```
So this will also skip for the default services controller for any services that are on UDN  networks right?

ricky-rav replied:
yes :)

trozet commented on go-controller/pkg/ovn/controller/services/services_controller.go:
```diff
@@ -544,16 +544,38 @@ func (c *Controller) RequestFullSync(nodeInfos []nodeInfo) {
+// belong to the network that this service controller is responsible for.
+func (c *Controller) skipService(name, namespace string) bool {
+	if util.IsNetworkSegmentationSupportEnabled() {
+		serviceNetwork, err := util.GetActiveNetworkForNamespace(namespace, c.nadLister)
```
so this GetActiveNetworkForNamespace iterates through all NADs in a for loop to try to find the one. This is going to be very slow with how many services we process. I guess its not too bad because its namespace scoped lookup, and there shouldn't be that many networks in a namespace. In the future lets make sure we have a github issue/jira or something to track making a controller to track networks mapping to namespaces.

trozet reviewed (commented):
Finished through 2nd commit. Will review the rest and test locally tmrw.

tssurya commented on go-controller/pkg/ovn/controller/services/node_tracker.go:
```diff
@@ -76,11 +78,13 @@ func (ni *nodeInfo) l3gatewayAddressesStr() []string {
 }
 
-func newNodeTracker(zone string, resyncFn func(nodes []nodeInfo)) *nodeTracker {
+func newNodeTracker(zone string, resyncFn func(nodes []nodeInfo), netInfo util.NetInfo) *nodeTracker {
```
do we need this really per network?
this tracker can just be global and used in each network right?

tssurya replied:
also the chassis template changes to SBDB per node; => if we do this per network; its going to be duplicated right?

ricky-rav replied:
There are parts of the node tracker that are relative to the network itself, like `gatewayRouterName` and `switchName`, which are used all throughout the code by the services controller to add LBs to the DB; there are other parts which are more relative to the node itself (`hostAddresses`, `l3gatewayAddresses`).
I agree that this can be further optimized, maybe in a follow up PR?

ricky-rav commented on go-controller/pkg/util/util.go:
```diff
@@ -506,6 +531,14 @@ func IsDefaultEndpointSlice(endpointSlice *discoveryv1.EndpointSlice) bool {
 }
 
+// IsEndpointSliceForNetwork checks if the provided EndpointSlice is meant for the given network
+func IsEndpointSliceForNetwork(endpointSlice *discoveryv1.EndpointSlice, network string) bool {
```
True, so let me rename it `IsMirroredEndpointSliceForNetwork`

trozet commented on test/e2e/service_udn.go:
```diff
@@ -0,0 +1,387 @@
+			// - a client in the UDN can reach the service through the node port on all host IPs of the node where the client runs
+			// - a client in the UDN on a different node than the backend pod can reach the service through the cluster IP
+			// - a client in the default network cannot reach the service in the UDN through the cluster IP
+			// - a client in the default network cannot reach the service in the UDN on any host IPs
```
we should add a test (maybe later if its not possible now) to make sure UDN pod cannot reach service in default network

ricky-rav replied:
I have it already, it's at the end, but I forgot to update the comment you highlighted :(

trozet commented:
I tested on a layer 3 UDN. Looks right. I see load balancers for my UDN service on GR and switch:

```
trozet@fedora:~/go/src/github.com/ovn-org/ovn-kubernetes/contrib$ oc get userdefinednetwork
NAME         AGE
l3-primary   3h20m


trozet@fedora:~/go/src/github.com/ovn-org/ovn-kubernetes/contrib$ oc get service
NAME          TYPE           CLUSTER-IP      EXTERNAL-IP    PORT(S)                       AGE
kubernetes    ClusterIP      10.96.0.1       <none>         443/TCP                       63m
my-service1   LoadBalancer   10.96.222.100   172.18.0.105   1337:31911/UDP,80:32093/TCP   46m



[root@ovn-worker2 ~]# ovn-nbctl lr-lb-list GR_default.l3.primary_ovn-worker2
UUID                                    LB                  PROTO      VIP                     IPs
9f8ace8b-586a-4e3c-b2e4-d21a1728688b    Service_default/    tcp        10.96.0.1:443           
dd29ae27-e006-4e17-ae10-714ddfdcf1a4    Service_default/    udp        ^NODEIP_IPv4_0:31911    10.20.1.3:1337
                                                            udp        ^NODEIP_IPv4_1:31911    10.20.1.3:1337
                                                            udp        ^NODEIP_IPv4_2:31911    10.20.1.3:1337
4aa1be56-9486-4e61-9ded-53ec7da079c3    Service_default/    tcp        ^NODEIP_IPv4_0:32093    10.20.1.3:80
                                                            tcp        ^NODEIP_IPv4_1:32093    10.20.1.3:80
                                                            tcp        ^NODEIP_IPv4_2:32093    10.20.1.3:80
129425af-87ee-4444-93ed-46d0916ae05c    Service_default/    udp        10.96.222.100:1337      10.20.1.3:1337
                                                            udp        172.18.0.105:1337       10.20.1.3:1337
58be4a6c-6df8-441b-ae4f-4a5ed105e2c0    Service_default/    tcp        10.96.222.100:80        10.20.1.3:80
                                                            tcp        172.18.0.105:80         10.20.1.3:80
[root@ovn-worker2 ~]# ovn-nbctl ls-lb-list default.l3.primary_ovn-worker2
UUID                                    LB                  PROTO      VIP                     IPs
9f8ace8b-586a-4e3c-b2e4-d21a1728688b    Service_default/    tcp        10.96.0.1:443           
dd29ae27-e006-4e17-ae10-714ddfdcf1a4    Service_default/    udp        ^NODEIP_IPv4_0:31911    10.20.1.3:1337
                                                            udp        ^NODEIP_IPv4_1:31911    10.20.1.3:1337
                                                            udp        ^NODEIP_IPv4_2:31911    10.20.1.3:1337
4aa1be56-9486-4e61-9ded-53ec7da079c3    Service_default/    tcp        ^NODEIP_IPv4_0:32093    10.20.1.3:80
                                                            tcp        ^NODEIP_IPv4_1:32093    10.20.1.3:80
                                                            tcp        ^NODEIP_IPv4_2:32093    10.20.1.3:80
129425af-87ee-4444-93ed-46d0916ae05c    Service_default/    udp        10.96.222.100:1337      10.20.1.3:1337
                                                            udp        172.18.0.105:1337       10.20.1.3:1337
58be4a6c-6df8-441b-ae4f-4a5ed105e2c0    Service_default/    tcp        10.96.222.100:80        10.20.1.3:80
                                                            tcp        172.18.0.105:80         10.20.1.3:80

```

trozet commented:
curl to the service from a pod across nodes also works

npinaeva commented on go-controller/pkg/ovn/controller/services/utils.go:
```diff
@@ -25,3 +31,19 @@ func IsHostEndpoint(endpointIP string) bool {
+	}
+
+	return map[string]string{
+		types.LoadBalancerOwnerExternalID: nsn.String(),
```
current external ids are a bit confusing to me.
We use the same lb.Name as before, which only depends on the service object and doesn't include anything network-specific, and it is a primary client index.
So let's say service's network change (I assume that is what happens when you add or delete primary UDN to the namespace?) then 2 controllers will try to update the same db entry (based on lb name), where old controller will try to clean it up and new controller will try to add it?
Also I didn't get if there is a way to have non-primary network service? Considering we store role in externalID, it should be the case, but I am not sure how it should work.

ricky-rav replied:
Ah are we allowing to move existing resources from one network to another? I thought it wasn't allowed. 

We're not launching the service controller from secondary networks, at least not yet :)

trozet replied:
we probably should include the network name as an external ID

npinaeva replied:
using proper external IDs would prevent such thing from happening, and we could use primaryID instead of name for client index.
that actually even should not be too difficult to implement, as lb.Name is not a server index.
Also if the network change for a service is not allowed, does it mean a service will always be attached to the network that was primary at the service creation time? Do we enforce this somehow?

ricky-rav replied:
wait wait, I thought that mine were proper external IDs. Are we talking about db indexes again?

npinaeva replied:
hehe yeah, but we can skip that part.
I think we may need to add network name to the lb.Name, and we need to explain what happens on namespace primary network change, and decide if we really need network role external id, considering network role is defined by the network name, and we don't support services for secondary networks.

trozet replied:
"Also if the network change for a service is not allowed, does it mean a service will always be attached to the network that was primary at the service creation time? Do we enforce this somehow?"

I think it should just transition to the new network if the primary network is created after the service. @ricky-rav lets make sure we test this scenario. Can be done in a follow up.

ricky-rav commented on go-controller/pkg/ovn/controller/services/services_controller.go:
```diff
@@ -650,21 +682,49 @@ func (c *Controller) queueServiceForEndpointSlice(endpointSlice *discovery.Endpo
 // serviceControllerKey returns a controller key for a Service but derived from
 // an EndpointSlice.
-func ServiceControllerKey(endpointSlice *discovery.EndpointSlice) (string, error) {
+// Not UDN-aware, is used for egress services
```
Yeah. I renamed it `GetServiceKeyFromEndpointSliceForDefaultNetwork` and I only call it from the egress service code. Once we add UDN support to egress services, we can make the function network aware. For now, since egress services don't support UDN, I didn't want to go as far as passing a `netInfo` from egress services to this util function.

ricky-rav commented on go-controller/pkg/ovn/controller/services/services_controller.go:
```diff
@@ -650,21 +682,49 @@ func (c *Controller) queueServiceForEndpointSlice(endpointSlice *discovery.Endpo
+	label := discovery.LabelServiceName
+	errTemplate := NoServiceLabelError
+	if !inDefaultNetwork {
+		label = types.LabelUserDefinedServiceName
```
yes :)

ricky-rav commented on go-controller/pkg/ovn/controller/services/utils.go:
```diff
@@ -25,3 +31,19 @@ func IsHostEndpoint(endpointIP string) bool {
+
+	role := types.NetworkRoleDefault
+	if !netInfo.IsDefault() {
+		role = types.NetworkRolePrimary
```
Actually yes, I can distinguish between default, primary and secondary, so that when we want to run the services controller from a secondary network, the code is ready.

ricky-rav commented:
In https://github.com/ovn-org/ovn-kubernetes/compare/14f2b26f6bc96cc030d5a333f92dad7d3831cc1c..60094957a5ad98eec60266086bf2247aff543c15 I addressed the remaining comments, mostly from Tim. 

I'll now add external IDs to logical routers and switches in the default network, as discussed internally, as well as a sync function at startup that adds them if not present.

ricky-rav commented:
The e2e's I have added now fail because dual stack node ports are broken on the services side. The endpoints are reachable when tested directly. The problem is that the LBs for the v6 node IP don't point to any backend pod:

```
$ ovn-nbctl ls-lb-list s9zmr_tenant.red_ovn-worker2
UUID                                    LB                  PROTO      VIP                     IPs
942b5805-4e0d-4ba9-ac3f-cb85484add2b    Service_services    udp        ^NODEIP_IPv6_0:32283    
4a6bcc0e-ed7a-494d-90b4-fd05a40c4303    Service_services    udp        10.96.166.193:80        10.128.0.3:80
729add49-70e8-478d-8fd9-af5813431c78    Service_services    udp        ^NODEIP_IPv4_0:32283    10.128.0.3:80
                                                            udp        ^NODEIP_IPv4_1:32283    10.128.0.3:80
```
I'll look into that tomorrow morning.

trozet commented on test/scripts/e2e-cp.sh:
```diff
@@ -154,7 +154,7 @@ fi
 # Only run network segmentation tests if they are explicitly requested
 NETWORK_SEGMENTATION_TESTS="Network Segmentation"
-if [ "${WHAT}" != "${NETWORK_SEGMENTATION_TESTS}" ]; then
+if [ "${WHAT}" != ${NETWORK_SEGMENTATION_TESTS}* ]; then
```
I think this is causing network segmentation tests to be run on non-network segmentation lanes

kyrtapz replied:
yeah, @ricky-rav I think you need double brackets:
```suggestion
if [[ "${WHAT}" != ${NETWORK_SEGMENTATION_TESTS}* ]]; then
```

ricky-rav commented:
The problem in https://github.com/ovn-org/ovn-kubernetes/pull/4555#issuecomment-2313233209 was simply due to the service in the e2e not being configured for dual stack.

In https://github.com/ovn-org/ovn-kubernetes/compare/6cc177f838218c28243aebd5f681aab66f225367..b215186322b3115e16920ec1bed766d73ceda204 I refactored the e2e test to make it easier to add more test cases.

TODO still:
- add external IDs to logical routers and switches in the default network, as discussed internally, as well as a sync function at startup that adds the external IDs if not present.
- have network-scoped LB names

dceara commented on test/e2e/network_segmentation_services.go:
```diff
@@ -0,0 +1,376 @@
+const OVNNodeHostCIDRs = "k8s.ovn.org/host-cidrs"
+
+// ParseNodeHostCIDRsDropNetMask returns the parsed host IP addresses found on a node's host CIDR annotation. Removes the mask.
+func ParseNodeHostCIDRsDropNetMask(node *kapi.Node) (sets.Set[string], error) {
+	addrAnnotation, ok := node.Annotations[OVNNodeHostCIDRs]
+	if !ok {
+		return nil, newAnnotationNotSetError("%s annotation not found for node %q", OVNNodeHostCIDRs, node.Name)
+	}
+
+	var cfg []string
+	if err := json.Unmarshal([]byte(addrAnnotation), &cfg); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal host cidrs annotation %s for node %q: %v",
+			addrAnnotation, node.Name, err)
+	}
+
+	for i, cidr := range cfg {
+		ip, _, err := net.ParseCIDR(cidr)
+		if err != nil || ip == nil {
+			return nil, fmt.Errorf("failed to parse node host cidr: %v", err)
+		}
+		cfg[i] = ip.String()
+	}
+	return sets.New(cfg...), nil
+}
```
This will also include the secondary node IPs (if multi-net is enabled - which is the case with UDN).  Connectiong NodePort services on remote node secondary IPs doesn't work.  That's the case for the default network too.

Would it make sense to use the `"k8s.ovn.org/node-primary-ifaddr"` annotation instead?  Something like (not fully tested):
```
// ParseNodeHostIPDropNetMask returns the parsed host IP addresses found on a node's host CIDR annotation. Removes the mask.
func ParseNodeHostIPDropNetMask(node *kapi.Node) (sets.Set[string], error) {
	nodeIfAddrAnnotation, ok := node.Annotations[OvnNodeIfAddr]
	if !ok {
		return nil, newAnnotationNotSetError("%s annotation not found for node %q", OvnNodeIfAddr, node.Name)
	}
	nodeIfAddr := &primaryIfAddrAnnotation{}
	if err := json.Unmarshal([]byte(nodeIfAddrAnnotation), nodeIfAddr); err != nil {
		return nil, fmt.Errorf("failed to unmarshal annotation: %s for node %q, err: %v", OvnNodeIfAddr, node.Name, err)
	}

	var cfg []string
	if nodeIfAddr.IPv4 != "" {
		cfg = append(cfg, nodeIfAddr.IPv4)
	}
	if nodeIfAddr.IPv6 != "" {
		cfg = append(cfg, nodeIfAddr.IPv6)
	}
	if len(cfg) == 0 {
		return nil, fmt.Errorf("node: %q does not have any IP information set", node.Name)
	}

	for i, cidr := range cfg {
		ip, _, err := net.ParseCIDR(cidr)
		if err != nil || ip == nil {
			return nil, fmt.Errorf("failed to parse node host cidr: %v", err)
		}
		cfg[i] = ip.String()
	}
	return sets.New(cfg...), nil
}
```

tssurya commented:
saving link: https://github.com/ovn-org/ovn-kubernetes/actions/runs/10616100554/job/29427820894?pr=4555

kyrtapz reviewed (approved):
The E2Es will need some minor changes but I am ok to do that as a followup. 
It can be a part of the L2 svc support.

ricky-rav commented:
We decided not to add network-related external IDs from OVN LBs on the default: https://github.com/ovn-org/ovn-kubernetes/compare/c7ed5ade6fcdba0cf36f073cfa3c2938a7968383..b8430ae864b98ddec90cde7ac3cdaa576a5fb2ea

We also decided to have network-scoped LB names, which allow a service to move from one network to another, since its old OVN LBs would get wiped by the previous network services controller and its new OVN LBs would get added to the new network services controller: https://github.com/ovn-org/ovn-kubernetes/compare/b8430ae864b98ddec90cde7ac3cdaa576a5fb2ea..3b136b6b9347dcb1c1f5118bd70d0a1ea41819c9

Lastly, we finalized the use cases to test in the e2e: https://github.com/ovn-org/ovn-kubernetes/compare/3b136b6b9347dcb1c1f5118bd70d0a1ea41819c9..b1ad639cad4060e4cbf4da3f2c55e228c00f7686

ricky-rav commented:
https://github.com/ovn-org/ovn-kubernetes/compare/b1ad639cad4060e4cbf4da3f2c55e228c00f7686..51652ba5ac4b344af2b59e0d1838a4ea37468d39 is just a rebase against master

trozet commented:
Just wondering about https://github.com/ovn-org/ovn-kubernetes/pull/4555/commits/a15c91ea8b347b982216c08c2c67d84aac5bf15d#r1736949256

ricky-rav commented on go-controller/pkg/ovn/secondary_layer3_network_controller.go:
```diff
@@ -537,6 +584,20 @@ func (oc *SecondaryLayer3NetworkController) Init(ctx context.Context) error {
+	// FIXME: When https://github.com/ovn-org/libovsdb/issues/235 is fixed,
+	// use IsTableSupported(nbdb.LoadBalancerGroup).
+	if _, _, err := util.RunOVNNbctl("--columns=_uuid", "list", "Load_Balancer_Group"); err != nil {
+		klog.Warningf("Load Balancer Group support enabled, however version of OVN in use does not support Load Balancer Groups.")
```
so... let's keep it as it is :D
